### PR TITLE
Add the STORE module take 2, for loaded certs, keys and more given a URI (cont)

### LINKED ISSUE #1958: STORE core library
For the [STORE project](https://github.com/openssl/openssl/projects/1), its core library (crypto/store/store_lib.c) needs the following functionlity:

- `STORE_load`, a function that loads objects the STORE library supports, given a URI.  It returns a `STACK_OF(STORE_INFO)` containing all the information it's been able to retrieve (a PEM file can contain quite a few items, for example).
- `STORE_INFO`, an opaque structure holding details on what has been loaded.
- functions to retrieve data of specific type from a `STORE_INFO` reference.
- functions to manipulate a `STORE_INFO`.
- functions to register and unregister loaders, which are handlers for each URI scheme that needs support.

### PR DESCRIPTION
**Pull request by levitte**
This is a new pull request to replace #2011, which has encountered github performance issues.

This STORE module adds the following functionality:

- A function STORE_open(), STORE_load() and STORE_close() that accesses
a URI and helps loading the supported objects (PKEYs, CERTs and CRLs
for the moment) from it.
- An opaque type STORE_INFO that holds information on each loaded object.
- A few functions to retrieve desired data from a STORE_INFO reference.
- Functions to register and unregister loaders for different URI schemes.  This enables dynamic addition of loaders from applications or from engines.

Also includes a loader for the "file" scheme.  The goal is to have it load PEM files and raw DER files alike, transparently.

Fixes #1958, #1959 

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by levitte** [comment]
Copying remaining big issues from #2011:

1) `STORE_` vs `OSSL_STORE_` prefix (not yet concluded as far as I can see)
2) ENDOFDATA vs `ferror` like interface.

I'm currently working on 3.  1 is waiting on team decision (I want it to become policy if we decide for `OSSL_STORE_`)...

**Comment by dwmw2** [comment]
We spoke of the desire to keep your test cases lifted from OpenConnect, in sync with the ones I have.

So I thought I should let you know I added a new one that you should incorporate:
http://git.infradead.org/users/dwmw2/openconnect.git/commitdiff/962d242

**Comment by levitte** [comment]
@dwmw2, hmmm, that will take some work to adapt, and especially since I try to be platform agnostic...

**Comment by levitte** [comment]
Of the listed things to fix, I realised I forgot to add one, @t-j-h's concern with opening up internal APIs, so I'll add that here:

1) `STORE_` vs `OSSL_STORE_` prefix (not yet concluded as far as I can see)
2) **[DONE]** ~ENDOFDATA vs~ `ferror` like interface.
3) Exposing `asn1_d2i_read_bio` as internal API vs making `ASN1_d2i_bio` accept a NULL `d2i` ( @t-j-h )

I feel unsure about (3)...  exposing things internally within libcrypto is already done for other things, I don't quite see why this one is a bad thing.  @t-j-h, are you very attached to an unreachable `asn1_d2i_read_bio` or can we let that item go?

**Comment by dwmw2** [comment]
> @dwmw2, hmmm, that will take some work to adapt, and especially since I try to be platform agnostic...

The test case itself, yeah. I need to find a Windows box (suspect Wine won't cut it here) and work that out. But having an expectation that it *should* work... that's a different story (qv).

**Comment by dwmw2** [comment]
Just attempting to work out what this looks like for PKCS#11... 

Given a RFC7512 URI string (which is actually a search term that may produce more than result), we call `STORE_open()` with that URI and then iterate over the results with `STORE_load()`, yes? 

I think that works sanely, and fits the same usage model as when there is more than one object within a given file. Will we eventually have a `SSL_CTX_use_STORE(uri, ui_method, ui_data, post_process, post_process_data)` function?

The only thing I think I'm missing is explicit documentation of the lifetime of the `ui_method` and `ui_data`. I need them to last as long as the last private key exported from the store. They can't go away when the `STORE_CTX` itself does. That's to deal with `CKA_ALWAYS_AUTHENTICATE` keys (and probably hotplug and other nastiness).

**Comment by levitte** [comment]
@dwmw2 wrote:
> Just attempting to work out what this looks like for PKCS#11...
> 
> Given a RFC7512 URI string (which is actually a search term that may produce more than result), we call `STORE_open()` with that URI and then iterate over the results with `STORE_load()`, yes?

Yes. Basically, it's roughly modeled from stdio, with the difference that we start with a URI in place of a file name.  

> I think that works sanely, and fits the same usage model as when there is more than one object within a given file. Will we eventually have a `SSL_CTX_use_STORE(uri, ui_method, ui_data, post_process, post_process_data)` function?

Obviously, you haven't seen #2696 and #2697... They're pretty rad, especially #2697. 

> The only thing I think I'm missing is explicit documentation of the lifetime of the `ui_method` and `ui_data`. I need them to last as long as the last private key exported from the store. They can't go away when the `STORE_CTX` itself does. That's to deal with `CKA_ALWAYS_AUTHENTICATE` keys (and probably hotplug and other nastiness).

I would expect `ui_method` to live for as long as the object (exe or dll) that implements it. As for `ui_data`, there's no general guarantee. Either way, the decision is up to the application or engine or whatever that implements the `ui_method`. 

I would expect that any condition specific to an engine *will be documented with said engine*

**Comment by levitte** [comment]
@dwmw2, and idea re `ui_method` and `ui_data` is to write a `UI_METHOD` of your own that wraps around the incoming `ui_method` and that caches all responses.  The next time there's an authentication request, simply replay the cache.  It's probably not the easiest `UI_METHOD` to write, but should be doable...

**Comment by levitte** [comment]
(previous comment removed...  I can't believe I wrote that out loud...)

**Comment by dwmw2** [comment]
So... If a pkcs11 STORE backend might exist one day and if my users might one day give me a string which happens to reference an object therein, as the cert they want to use... Then I need to have read that documentation before I write my application today?

I would be much happier with lifetime rules for the ui_data warning up front that *some* back ends may keep references to it for as long as keys exist.

Those other PRs... Neat. Need to think about how we handle that for system trust, especially w.r.t. `find_issuer()` which will want to specify the subject in the uri and *also* postprocess.

**Comment by levitte** [comment]
Do you mean the `find_issuer` in `crypto/x509/x509_vfy.c`?  In that case, I've no idea what you're talking about...  However, if you're thinking in more general searching terms, you might want to have a look at the add-on in #2688.

**Comment by levitte** [comment]
> So... If a pkcs11 STORE backend might exist one day and if my users might one day give me a string which happens to reference an object therein, as the cert they want to use... Then I need to have read that documentation before I write my application today?

Well, it would seem appropriate to read docs that come with the backends you're about to use, if there is any, no?  I don't understand why that would be questionable...  I find it as natural as having to read `socket(7)` to figure out the socket level option names and arguments for `getsockopt(3)`...

> I would be much happier with lifetime rules for the ui_data warning up front that *some* back ends may keep references to it for as long as keys exist.

Mm, good point.  I think I can write something generic enough.  I'll get back to that on Monday.

**Comment by dwmw2** [comment]
Yes I'm thinking of that `find_issuer()`. Linux distributions now tend to have trust roots in a PKCS#11 token and only export it to a flat file for backward compatibility. Iterating over all  objects​ in the token *isn't* the best way to do it. I'll take a closer look...

**Comment by dwmw2** [comment]
For the other thing... I write my application to use the STORE interface long before the back end in question even exists. A sysadmin installs the PKCS#11 back end once it's written, and a user chooses a cert with a PKCS#11 URI.   It wasn't "me" who used that back end.

**Comment by levitte** [comment]
Ah, I get your point.

The *normal* thing, at least from the usual OpenSSL perspective, would be to have the backend duplicate the UI data...  which is, of course, difficult with the current UI API.  But then, I could easily add something that would do exactly that.  I would think that's a saner way of doing things.

**Comment by dwmw2** [comment]
"duplicate the ui_data"?

This is an opaque (to OpenSSL) datum which ends up being passed back to the application in order to request the password. It contains GUI context or something, and may have a chain of other refcounted objects  hanging off it that OpenSSL can know nothing about....

**Comment by levitte** [comment]
See #3575.

**Comment by levitte** [comment]
My thought is that if the ui_data can't be duplicated (regardless of reason), then your backend can't support things like `CKA_ALWAYS_AUTHENTICATE`.  The benefit of doing it this way is expliciteness.  The application will have to *explicitely* support this kind of duplication, and the backend will be able to return an error if that support is missing, which I think is much better than merely hoping that the application is well behaved and keeps the ui_data around for its life time (and getting mysterious crashes if it doesn't).

**Comment by dwmw2** [comment]
Ack, that works. Thanks.  A nice obvious failure mode if an app forgets to provide a dup method -- even today with just the file backend where it isn't strictly  needed -- is extra helpful.

**Comment by dwmw2** [comment]
I'm starting to wonder if a callback function to get the password might be better than a UI. People build without UI support (e.g. MacOS) and do we want it to be a dependency? The "user" prompts we offer are never going to be perfectly suited for presenting directly to the user anyway (with translations and referring to the object being unlockedj with the terminology that the app+user use for it.

Using a callback also lets us punt on the charset question and insist we only take UTF-8. Because now there is *app* code between the STORE and the `UI_OpenSSL()` *if* they happen to use that, so we can tell them it's their responsibility to convert.

We also get to convey other flags like "wrong PIN" and "last chance" with our own callback, in machine readable way (i.e. so the app can modify its behaviour and stop using the cached pin, etc.)

**Comment by levitte** [comment]
I disagree.  Having a "simple" callback function will require that the application knows exactly what the backend will want to prompt for, or that the backend might find itself limited to what the callback function takes.

**Comment by dwmw2** [comment]
Yeah, that's true..  at least for a very simple function, and once you've d made it less simple it probably looks a bit like a UI... :)

Apps can put a lot of the context information into the ui_data, which helps with my first point. I'm happy to say you Sam well ought to handle charsets properly, which resolved my second point.

What remains is the "you got it wrong last time" and "you are screwed if you get it wrong again" and other time-of-invocation flags. Those are relatively easy to add to the existing UI I suppose.

**Comment by dwmw2** [comment]
Damn autocorrect

**Comment by dwmw2** [comment]
So #3575 gives us the management of `ui_data` that allows this to work with keys that require a PIN every time. What remains on this side is to make it abundantly clear to applications using the STORE interface that they must expect such, and not expect that the `ui_data` goes out of scope the moment they're done with the STORE. Bonus points for making it hard to get it wrong, by *requiring* it even for files somehow.

**Comment by levitte** [comment]
I'm sorry, I didn't quite understand that last part.  "*requiring* it even for files somehow"???  How did I do that?

**Comment by dwmw2** [comment]
You didn't, and I'm not sure how to offhand. But if you can make it hard for application authors to get this wrong, I'll award bonus points :)

**Comment by levitte** [comment]
I'm not even sure what "getting it wrong" would mean in this context, especially re the file scheme.  Let's just not make things overly complicated, shall we?

**Comment by dwmw2** [comment]
*"getting it wrong"* for an application would involve assuming your user will only ever use file URIs and not bothering to implement the mechanism provided in #3575, so you get unexpected breakage when the user happens to choose a hardware key that always needs a PIN.

Forcing the application to get it right... well, as I said I can't think of a great way of doing that. You're right that it's probably overly complicated. Maybe we could just make sure that the file loader invokes the UI data_destructor method only when the last EVP_PKEY goes away, not when the STORE itself is (earlier) destroyed. Or maybe a `$OPENSSL_STORE_ALWAYS_PROMPT` to make it ask for the file password again every time a key is used, even though it didn't need to. But I'm not really advocating either of those; they're just examples of the class of things we *might* consider doing. If we could find one that was simple and effective. But I suspect that documenting it extremely clearly is the best we can do.

**Comment by levitte** [comment]
The file scheme loader *has no interest* in the UI user data after the loaded keys have been unlocked.  *None*.  I don't see why it should.  I understand your desire to set an example, but really...

**Comment by levitte** [comment]
I renamed the lot by changing the `STORE` prefix to `OSSLSTORE`.  That should make it unique enough without becoming too bulky and without causing other trouble in our subsystems (the error codes aren't allowed to have an underscore in the prefix, for example).

Now, to rebase it on a fresher master

**Comment by levitte** [comment]
Update:

1) **[DONE]** ~`STORE_` vs `OSSL_STORE_` prefix (not yet concluded as far as I can see)~ (the prefix became `OSSLSTORE_`)
2) **[DONE]** ~ENDOFDATA vs~ `ferror` like interface.
3) Exposing `asn1_d2i_read_bio` as internal API vs making `ASN1_d2i_bio` accept a NULL `d2i` ( @t-j-h )

I've received no response whatsoever re 3), so I think I'll ignore it.  After all, it's internal to libcrypto, we have a number of similar exposures already, I really don't see much harm.

@dwmw2, I've heard your concerns re character encoding, and I agree that we need a general overhaul on the matter.  *This* PR is not the place.

@openssl/committers and @openssl/omc, I think it's time for another review, maybe even a final one...

**Comment by mattcaswell** [comment]
Re: STORE_ vs OSSL_STORE_ vs OSSLSTORE_, I think there was a comment from @t-j-h  (I think via email) along the lines that it should be OSSLSTORE and we should fix "make errors", but I don't recall exactly. Has that been responded to?

**Comment by levitte** [comment]
"fixing" mkerr.pl to accept underscores in module names has ramifications that I have no desire digging into.  The fact that the module can't include an underscore makes for a very clear distinction where the module name ends and where the function / reason indication starts.

So, instead of battling what double prefix to use, I chose a simple rename.  The initial request was to make it unique enough, and this rename does solve that problem.  It also avoids getting extremely long symbol names (they are already long enough, adding something like `OPENSSL_` in front of them became bulky, just like you commented by email).

I certainly hope we can stay with this compromise.

File: doc/man1/storeutl.pod
Comment (mattcaswell) [inline doc/man1/storeutl.pod]: Seems to be missing documentation for the -text option.

File: apps/storeutl.c
Comment (mattcaswell) [inline apps/storeutl.c]: Should this be an OPENSSL_assert() here? This isn't actually an OpenSSL bug (necessarily), but an engine bug (which may be external).

File: crypto/store/loader_file.c
Comment (mattcaswell) [inline crypto/store/loader_file.c]: {} not required

File: crypto/store/loader_file.c
Comment (mattcaswell) [inline crypto/store/loader_file.c]: Unecessary {}

File: crypto/store/loader_file.c
Comment (mattcaswell) [inline crypto/store/loader_file.c]: It looks to me like 10 is the exact length of "Proc-Type:" that PEM_get_EVP_CIPHER_INFO() is expecting to see in the header. That function will fail if it doesn't see that (unless the header length is 0 or just contains "\n".

File: crypto/store/store_register.c
Comment (mattcaswell) [inline crypto/store/store_register.c]: return NULL?

File: doc/man7/store.pod
Comment (mattcaswell) [inline doc/man7/store.pod]: Probably all the above library man pages should reference this page somewhere prominent near the beginning of the page (and in the SEE ALSO section).

**Review by mattcaswell — CHANGES_REQUESTED** [review]
Some relatively minor comments. When I approve this I will do so subject to no objections being raised from @t-j-h

File: doc/man1/storeutl.pod
Comment (levitte) [inline doc/man1/storeutl.pod]: Ah, thanks.  Fixed

File: apps/storeutl.c
Comment (levitte) [inline apps/storeutl.c]: Mm...  considering this is an app, and that the correct response is to break out, I wasn't too concerned.  But ok, it's probably a nice thing to allow the store to close either way, so I'll just print an error and break out of the loop.

File: crypto/store/loader_file.c
Comment (levitte) [inline crypto/store/loader_file.c]: Good catch, I'll change the comment

File: crypto/store/loader_file.c
Comment (levitte) [inline crypto/store/loader_file.c]: Fixed

File: crypto/store/loader_file.c
Comment (levitte) [inline crypto/store/loader_file.c]: Fixed

File: doc/man7/store.pod
Comment (levitte) [inline doc/man7/store.pod]: This file should also be renamed to `osslstore.pod`, should it not?

Comment (mattcaswell) [inline doc/man7/store.pod]: Yes, makes sense.

Comment (richsalz) [inline doc/man7/store.pod]: If you changed it to "osslstore - Store retrieval functions" then yes.  The filename should match ojne of the entries in the NAME section.  The find-doc-nits script should have complained.

**Comment by levitte** [comment]
I think I've addressed all outstanding change requests

**Review by mattcaswell — APPROVED** [review]
Approved subject to @t-j-h not objecting WRT the naming issue. Please check with him before merging.

**Comment by levitte** [comment]
(for our external readers)
We've had an internal debate, and we ended up deciding on claiming the name spaces OSSL and OPENSSL, using them as prefixes with a separating underscore.  #3781 was needed to fulfill this.

So we finally end up with a module called OPENSSL_STORE

File: util/private.num
Comment (mattcaswell) [inline util/private.num]: These look out of date

**Review by mattcaswell — APPROVED** [review]
Subject to the one nit being fixed.

File: util/private.num
Comment (levitte) [inline util/private.num]: Good catch!  Fixed.

**Comment by levitte** [comment]
I found that the file loader might not correctly detect if a piece of data didn't match any object handler or if some other error occured in the object handler that's currently being attempted.  The difference is crucial to see if it should continue to check every remaining object handler or not.  Also, there are cases that I discovered while working with another branch, where the STORE error messages kind of shadow the issues found with a piece of data.

I'm currently tinkering with the file loader to fix these issues.

**Comment by levitte** [comment]
Issues fixed, it's all in the last commit.  The others have the name changed integrated in them.

**Review by mattcaswell — APPROVED** [review]
One minor nit. Approved assuming that's fixed.

File: crypto/store/loader_file.c
Comment (levitte) [inline crypto/store/loader_file.c]: Fixed

**Comment by levitte** [comment]
Apart from the comment fix, I found a nasty little bug that had escaped Travis because I didn't have it run the full set of tests.  All in the just added commit, please have a look

**Comment by levitte** [comment]
(I've set it to run extended tests, I certainly hope that will be the last thing to deal with)

**Review by mattcaswell — APPROVED** [review]
Approved assuming the builds pass.

**Comment by levitte** [comment]
The job to watch is the `enable-msan` one...  that's where I found failures in my personal Travis instance

**Comment by levitte** [comment]
Yes, builds passed!  Time to merge.

**Comment by levitte** [comment]
Merged.  c785fd48e68611c837f4e30027c02b84525501af to 6fc1d33c90015d3ad5738ec99aaa12fdb9640295

Thanks for the review, everyone

**Comment by levitte** [comment]
Next in line, I propose starting with a couple of easy ones, #3483 and #2745.  After that, #2746 (which might or might not be controversial)

**Comment by levitte** [comment]
Another choice might be to have a look at #2688, which is a requirement for #3481 (the CNG engine, that I know has caught some interest).

File: crypto/store/store_register.c
Comment (kaduk) [inline crypto/store/store_register.c]: I thought we super-double-plus ensured that the scheme field was never NULL, so most of this is redundant.

File: crypto/store/store_lib.c
Comment (kaduk) [inline crypto/store/store_lib.c]: These functions seem like a giant footgun, as we don't enforce in OSSL_STORE_register_loader() that all the method functions are non-NULL.
Maybe I missed some previous discussion on this pull request and this is in fact the behavior we want from our "don't check user input for NULL" policy, but my personal preference would be to reject registering an incomplete loader.

File: crypto/store/store_register.c
Comment (levitte) [inline crypto/store/store_register.c]: Good point.

File: crypto/store/store_lib.c
Comment (levitte) [inline crypto/store/store_lib.c]: Good point.

**Comment by levitte** [comment]
@kaduk, see #3805

File: crypto/store/loader_file.c
Comment (kaduk) [inline crypto/store/loader_file.c]: Do things break if this is pem_check_suffix() instead of strcmp()?

File: crypto/store/loader_file.c
Comment (kaduk) [inline crypto/store/loader_file.c]: No one complained about magic numbers in pointer arithmetic? ;)

File: crypto/store/loader_file.c
Comment (bernd-edlinger) [inline crypto/store/loader_file.c]: by the way, isn't the DNS part of an URI case-insensitive?

File: crypto/store/loader_file.c
Comment (levitte) [inline crypto/store/loader_file.c]: Yes.  `PEM_STRING_PUBLIC` identifies PKCS#8 public keys, which has the algo OID as part of the contents.  As far as I know, that's what we actively support.  Is there are reason you'd like us to support public keys that are typed in the PEM name?

File: crypto/store/loader_file.c
Comment (levitte) [inline crypto/store/loader_file.c]: Obviously not...  but then, considering the comparison with "file:" just above, this isn't much of a mystery, is it?

File: crypto/store/loader_file.c
Comment (levitte) [inline crypto/store/loader_file.c]: Good point.

Comment (levitte) [inline crypto/store/loader_file.c]: #3827

File: crypto/store/loader_file.c
Comment (kaduk) [inline crypto/store/loader_file.c]: No; I wasn't quite sure on the difference between the forms, and noted that we have a couple examples of "RSA PUBLIC KEY" in documentation, so wondered what the difference was.  Thanks for clarifying.

File: doc/man7/ossl_store.pod
Comment (Et7f3) [inline doc/man7/ossl_store.pod:66]: I have issue with this example provided. The first time I saw this function it hasn't this prototype: https://github.com/openssl/openssl/blame/4fd397821139723fd4e51a03e92df33e9a9fadcc/include/openssl/store.h#L55-L58 I will use code from storeutl to help me.

**Pull request closed without merging**
